Add WORKFLOW.md (#25461)
grafana/grafana: WORKFLOW.md

# Grafana workflow

This document is based on [GOVERNANCE.md](GOVERNANCE.md). We assume good faith and intend to keep all processes as lightweight as possible but as specific as required. In case of disagreements about anything in this document, GOVERNANCE.md applies.

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC2119](http://tools.ietf.org/html/rfc2119).

Git and [GitHub terminology](https://help.github.com/en/github/getting-started-with-github/github-glossary) are used throughout this document.

> Team members and their access to repositories is maintained through [GitHub teams](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/about-teams). Team maintainers add and remove team members as per GOVERNANCE.md

Reviewer: ```suggestion
Team members and their access to repositories is maintained through [GitHub teams](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/about-teams). Team maintainers add and remove team members as outlined in GOVERNANCE.md.
```

> # Code changes

Reviewer: ```suggestion
## Code changes
```

Reviewer: Please remove the empty line above this paragraph.

Reviewer: Adjust the heading levels that follow this to be +1 in depth until the next h1 to h2 adjustment.

Author: Newline removed, heading is how I think the structure makes sense.

## Proposing changes

> Examples of proposed changes are overarching architecture, subsets, and specific code/graphical elements. Proposed changes SHOULD cover the big picture and intention, but individual parts SHOULD be split into the smallest possible changes. Changes SHOULD be based on and target the master branch. Depending on size of the proposed change, they SHOULD be discussed, in increasing order of change size and complexity:

Reviewer: ```suggestion
Examples of proposed changes are overarching architecture, subsets, and specific code or graphical elements. Proposed changes SHOULD cover the big picture and intention, but individual parts SHOULD be split into the smallest possible changes. Changes SHOULD be based on and target the master branch. Depending on size of the proposed change, each change SHOULD be discussed, in increasing order of change size and complexity:
```

> Directly in a Pull Request - this MAY be done, but SHOULD not be the common case

Reviewer: ```suggestion
Directly in a pull request (PR) - This MAY be done, but SHOULD not be the common case.
```

Reviewer: This list SHOULD be an unordered list.

Issue
Developer mailing list
Design document, shared via Google Docs, accessible to at least all team members

Significant changes MUST be discussed and agreed upon with the relevant subsystem maintainers.

> ## Merging Pull Requests

Reviewer: ```suggestion
### Merging PRs
```

> Depending on the size and complexity of a Pull Request, different requirements MUST be applied. Any team member contributing substantially to a Pull Request MUST NOT count against review requirements.

Reviewer: ```suggestion

Depending on the size and complexity of a PR, different requirements MUST be applied. Any team member contributing substantially to a PR MUST NOT count against review requirements.
```

Reviewer: Fix the content in lines 26-33 so that they make a coherent paragraph or two. Also, end the sentences with periods.

Commits MUST be merged into master using Pull Requests. They MUST NOT be merged into master directly.
Every merge MUST be approved by at least one team member
Non-trivial changes MUST be approved by at least
two team members, or
one subsystem maintainer
Significant changes MUST be approved by at least
two team members, AND
the relevant subsystem maintainer

> Pull Requests MUST be [reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests) and [approved](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/approving-a-pull-request-with-required-reviews) via GitHub’s review system. Reviewers MAY write comments if approving and MUST write comments if rejecting a Pull Request or if requesting changes.

Reviewer: ```suggestion
PRs MUST be [reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests) and [approved](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/approving-a-pull-request-with-required-reviews) via GitHub’s review system. 

- Reviewers MAY write comments if approving. 
- Reviewers MUST write comments if rejecting a PR or if requesting changes.
```

> Once a Pull Request is approved as per above, any team member MAY merge the Pull Request.

Reviewer: ```suggestion
Once a PR is approved as per above, any team member MAY merge the PR.
```

> # Release workflow

Reviewer: ```suggestion
## Release workflow
```

Reviewer: Delete the empty line above and adjust the heading levels that follow to be +1 depth.

> ## Branch structure

Reviewer: Add an empty line after each heading to increase readability for maintainers.

Grafana uses trunk-based development.

> In particular, we found that the following principles match how we work

Reviewer: Formatting is unclear. It looks like the following lines should be an unordered list. If so, please add the necessary formatting and end line 44 with a colon.

Master and release branches MUST always build without failure.
Branches SHOULD be merged often. Larger changes SHOULD be activated with feature flags until they are ready. Long-lived development branches SHOULD be avoided.
Changes MAY be enabled by default once they are in a complete state
Changes which span multiple Pull Requests MUST be described in an overarching issue or Google Doc.

## Releases
Releases MUST follow [Semantic Versioning](https://semver.org/) in naming and SHOULD follow Semantic Versioning as closely as reasonably possible for non-library software.

> Release branches MUST be split from the following branches.

Reviewer: ```suggestion
Release branches MUST be split from the following branches:
```

Reviewer: Make the following three lines an unordered list.

MAJOR release branches MUST be based on master.
MINOR release branches MUST be based on master.
PATCH release branches MUST be split from the relevant MINOR release branch’s most current PATCH

> Security releases follow the same process but MUST be prepared in secret. Security releases MUST NOT include changes which are not related to the security fix. Normal release processes MUST accommodate the security release process. SECURITY.md MUST be followed.

Reviewer: Is there a reason SECURITY.md was not linked? Serious question, not sarcastic. If not, then please link.

If there is, then ignore this comment.

Author: Getting SECURITY.md in place in WIP, but I didn't want to block on that.

> Pull requests intended for inclusion in the next PATCH release MUST be labeled with `cherry-pick-needed` so they can be picked up by automated release tooling.

Reviewer: ```suggestion
PRs intended for inclusion in the next PATCH release MUST be labeled with `cherry-pick-needed` so that they can be picked up by automated release tooling.
```

Author: Adapted.

> Releases follow the following cadence

Reviewer: Apply unordered list formatting to the following three lines.

MAJOR: Yearly
MINOR: Every 4-6 weeks
PATCH: As needed

Releases SHOULD NOT be delayed by pending changes.

> Releases MUST be coordinated with the relevant subsystem maintainers.

Reviewer: In this context, what is a "relevant subsystem maintainer"? Consider listing a few examples.

Author: I can refer to https://github.com/grafana/grafana/blob/master/MAINTAINERS.md , though that's already done in governance, and we refer to governance in its entirety.

# Grafana workflow

This document is based on [GOVERNANCE.md](GOVERNANCE.md). We assume good faith and intend to keep all processes as lightweight as possible but as specific as required. In case of disagreements about anything in this document, GOVERNANCE.md applies.

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC2119](http://tools.ietf.org/html/rfc2119).

Git and [GitHub terminology](https://help.github.com/en/github/getting-started-with-github/github-glossary) are used throughout this document.

Team members and their access to repositories is maintained through [GitHub teams](https://help.github.com/en/github/setting-up-and-managing-organizations-and-teams/about-teams). Team maintainers add and remove team members as outlined in GOVERNANCE.md.

# Code changes

## Proposing changes

> Examples of proposed changes are overarching architecture, subsets, and specific code or graphical elements. Proposed changes SHOULD cover the big picture and intention, but individual parts SHOULD be split into the smallest possible changes. Changes SHOULD be based on and target the master branch. Depending on size of the proposed change, each change SHOULD be discussed, in increasing order of change size and complexity:

Author: Fixed in new commit.

  * two team members, AND
  * the relevant subsystem maintainer

Pull Requests MUST be [reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests) and [approved](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/approving-a-pull-request-with-required-reviews) via GitHub’s review system.
* Reviewers MAY write comments if approving
* Reviewers MUST write comments if rejecting a Pull Request or if requesting changes.

Once a Pull Request is approved as per above, any team member MAY merge the Pull Request.

# Release workflow

## Branch structure

Grafana uses trunk-based development.

In particular, we found that the following principles match how we work
* Master and release branches MUST always build without failure.
* Branches SHOULD be merged often. Larger changes SHOULD be activated with feature flags until they are ready. Long-lived development branches SHOULD be avoided.
* Changes MAY be enabled by default once they are in a complete state
* Changes which span multiple Pull Requests MUST be described in an overarching issue or Google Doc.

## Releases

> Releases MUST follow [Semantic Versioning](https://semver.org/) in naming and SHOULD follow Semantic Versioning as closely as reasonably possible for non-library software.

Reviewer 2: That's why it says `MUST` and `SHOULD`
